### PR TITLE
less memory allocation in GATKRead operations

### PR DESCRIPTION
**Pull request by akiezun**
reduced memory consumption: identified by profiling HC. 
+ added a bunch of tests while I was there.

note: the equals/hashcode changes are just for clean code - no perf improvements

@lbergelson can you have a look?

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java [hunk 1]
```diff
@@ -269,6 +269,18 @@ default byte getBaseQuality(final int i){
     }
 
     /**
+     * Return the cigar element at a given index.
+     *
+     * Note: the default implementation return <code>getCigarElements().get(i)</code>.
+     * Subclasses may override, for example to reduce the memory allocation or improve speed.
+     * @throws IndexOutOfBoundsException if the index is out of range  (<tt>index &lt; 0 || index &gt;= numCigarElements()</tt>)
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java]: What's with this bizare &gt?

File: src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java [hunk 2]
```diff
@@ -269,6 +269,18 @@ default byte getBaseQuality(final int i){
...
+     * @throws
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java]: throws...?

File: src/main/java/org/broadinstitute/hellbender/utils/read/SAMRecordToGATKReadAdapter.java [hunk 3]
```diff
@@ -579,13 +593,13 @@ public boolean equals(Object o) {
...
 
-        return !(samRecord != null ? !samRecord.equals(that.samRecord) : that.samRecord != null);
+        return Objects.equals(samRecord, that.samRecord);
 
     }
 
     @Override
     public int hashCode() {
-        return samRecord != null ? samRecord.hashCode() : 0;
+        return Objects.hashCode(samRecord);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/read/SAMRecordToGATKReadAdapter.java]: That's nice, didn't know about those...

**Comment by lbergelson** [comment]
@akiezun Looks like some html escaping snuck into the documentation.  +1 otherwise.

File: src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java]: javadocs are html to this is html for `<` and `>`

File: src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java]: fixed

**Pull request merged**